### PR TITLE
Wiring up CAPZ bootstrapping extension for Windows

### PR DESCRIPTION
**Pull request by marosset**
Signed-off-by: Mark Rossetti <marosset@microsoft.com>

 

 
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Adds an Azure VM extension to ensure VM bootstrapping succeeds for Windows nodes.
```

### CODE REVIEW

File: api/v1alpha3/zz_generated.conversion.go [hunk 1]
```diff
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated_core_v1alpha3
```
Comment (CecileRobertMichon) [inline api/v1alpha3/zz_generated.conversion.go:1]: @marosset did you use go 1.17 to run make generate? If so can you please run it again with 1.16? The project still expected 1.16

Comment (marosset) [reply api/v1alpha3/zz_generated.conversion.go:1]: Let me doublecheck and update the generated files.

Comment (marosset) [reply api/v1alpha3/zz_generated.conversion.go]: I re-ran `make generate` with go 1.16.8 and pushed the changes.

**Comment by jsturtevant** [comment]
looks like the [extension failed](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1734/pull-cluster-api-provider-azure-e2e-windows/1442518911416274944/artifacts/clusters/capz-e2e-xqm5a5-win-ha/azure-activity-logs/capz-e2e-xqm5a5-win-ha.log) on the Windows job:

```
        "message": "Microsoft.Compute/virtualMachines/extensions/write",
        "statusMessage": "{\"status\":\"Failed\",\"error\":{\"code\":\"ResourceOperationFailure\",\"message\":\"The resource operation completed with terminal provisioning state 'Failed'.\",\"details\":[{\"code\":\"VMExtensionProvisioningError\",\"message\":\"VM has reported a failure when processing extension 'CAPZ.Windows.Bootstrapping'. Error message: \\\"Command execution finished, but failed because it returned a non-zero exit code of: '-1'\\\"\\r\\n\\r\\nMore information on troubleshooting is available at https://aka.ms/vmextensionwindowstroubleshoot \"}]}}"
    },
```

**Comment by marosset** [comment]
> looks like the [extension failed](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1734/pull-cluster-api-provider-azure-e2e-windows/1442518911416274944/artifacts/clusters/capz-e2e-xqm5a5-win-ha/azure-activity-logs/capz-e2e-xqm5a5-win-ha.log) on the Windows job:
> 
> ```
>         "message": "Microsoft.Compute/virtualMachines/extensions/write",
>         "statusMessage": "{\"status\":\"Failed\",\"error\":{\"code\":\"ResourceOperationFailure\",\"message\":\"The resource operation completed with terminal provisioning state 'Failed'.\",\"details\":[{\"code\":\"VMExtensionProvisioningError\",\"message\":\"VM has reported a failure when processing extension 'CAPZ.Windows.Bootstrapping'. Error message: \\\"Command execution finished, but failed because it returned a non-zero exit code of: '-1'\\\"\\r\\n\\r\\nMore information on troubleshooting is available at https://aka.ms/vmextensionwindowstroubleshoot \"}]}}"
>     },
> ```

Let me run this locally again and see if it fails for me.
Also I will update the return code to `-2` (ERROR_FILE_NOT_FOUND) which is more appropriate.

File: azure/defaults.go [hunk 2]
```diff
@@ -299,21 +309,37 @@ func GetDefaultWindowsImage(k8sVersion string) (*infrav1.Image, error) {
...
+			Name:      "CAPZ.Linux.Bootstrapping",
+			VMName:    vmName,
+			Publisher: "Microsoft.Azure.ContainerUpstream",
+			Version:   "1.0",
+			ProtectedSettings: map[string]string{
+				"commandToExecute": LinuxBootstrapExtensionCommand,
+			},
+		}
+	} else if osType == WindowsOS && cloud == azure.PublicCloud.Name {
+		// This command always returns success (for now).
```
Comment (jsturtevant) [inline azure/defaults.go:330]: It should fail with the powershell if the command isn't there?

**Review by jsturtevant — COMMENTED** [review]
One minor comment that I think is left over. Otherwise looks good. It is a bit awkward to call the extension property `VMName` when it is a scale set but I don't have a good alternative.

File: azure/defaults.go (hunk 2 above)
Comment (marosset) [reply azure/defaults.go:330]: good catch - this comment was stay (initially I was testing with `cmd.exe /c exit /b 0` for the extension command which did always return 0/success)

Comment (marosset) [reply azure/defaults.go]: Fixed

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @devigned @jsturtevant

**Review by devigned — APPROVED** [review]
/lgtm

**Comment by devigned** [comment]
/test pull-cluster-api-provider-azure-e2e-windows

LB flake. Let's give another run.

**Pull request merged**